pinot-core/src/main/java/org/apache/pinot/core/operator/streaming/BaseStreamBlockCombineOperator.java

```diff
@@ -0,0 +1,176 @@
+import org.apache.pinot.core.operator.blocks.results.MetadataResultsBlock;
+import org.apache.pinot.core.operator.combine.BaseCombineOperator;
+import org.apache.pinot.core.operator.combine.CombineOperatorUtils;
+import org.apache.pinot.core.operator.combine.merger.ResultBlockMerger;
+import org.apache.pinot.core.query.request.context.QueryContext;
+import org.apache.pinot.core.query.scheduler.resources.ResourceManager;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+
+public abstract class BaseStreamBlockCombineOperator<T extends BaseResultsBlock>
+    extends BaseCombineOperator<BaseResultsBlock> {
+  private static final Logger LOGGER = LoggerFactory.getLogger(BaseStreamBlockCombineOperator.class);
+
+  /**
+   * Special results block to indicate that this is the last results block for a child operator in the list
+   */
+  public static final MetadataResultsBlock LAST_RESULTS_BLOCK = new MetadataResultsBlock();
+
+  // Use a BlockingQueue to store the intermediate results blocks
+  protected final BlockingQueue<BaseResultsBlock> _blockingQueue = new LinkedBlockingQueue<>();
+  protected final ResultBlockMerger<T> _combineFunction;
+
+  protected int _numOperatorsFinished;
+  protected BaseResultsBlock _exceptionBlock;
+
+  public BaseStreamBlockCombineOperator(ResultBlockMerger<T> combineFunction, List<Operator> operators,
+      QueryContext queryContext, ExecutorService executorService) {
+    super(operators, queryContext, executorService);
+    _combineFunction = combineFunction;
+    _numOperatorsFinished = 0;
+    _exceptionBlock = null;
+  }
+
+  @Override
+  protected BaseResultsBlock getNextBlock() {
+    long endTimeMs = _queryContext.getEndTimeMs();
+    while (_exceptionBlock == null && _numOperatorsFinished < _numOperators) {
+      try {
+        BaseResultsBlock resultsBlock =
+            _blockingQueue.poll(endTimeMs - System.currentTimeMillis(), TimeUnit.MILLISECONDS);
+
+        if (resultsBlock == null) {
+          // Query times out, skip streaming the remaining results blocks
+          LOGGER.error("Timed out while polling results block (query: {})", _queryContext);
+          _exceptionBlock = new ExceptionResultsBlock(QueryException.getException(
+              QueryException.EXECUTION_TIMEOUT_ERROR, new TimeoutException("Timed out while polling results block")));
+          return _exceptionBlock;
+        }
+        if (resultsBlock.getProcessingExceptions() != null) {
+          // Caught exception while processing segment, skip streaming the remaining results blocks and directly return
+          // the exception
+          _exceptionBlock = resultsBlock;
+          return _exceptionBlock;
+        }
+        if (resultsBlock == LAST_RESULTS_BLOCK) {
+          // Caught LAST_RESULTS_BLOCK from a specific task, indicated it has finished.
+          // Skip returning this metadata block and continue to process the next from the _blockingQueue.
+          _numOperatorsFinished++;
```

Comment:
I don't follow this part of the logic. Why do we increase this counter when hitting last block?